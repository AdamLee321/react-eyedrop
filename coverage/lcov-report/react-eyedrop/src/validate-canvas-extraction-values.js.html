<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for react-eyedrop/src/validate-canvas-extraction-values.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">react-eyedrop/src</a> validate-canvas-extraction-values.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>12/12</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>8/8</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>1/1</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>8/8</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">// @flow
&nbsp;
type ExtractionValues = {
  x: number,
  y: number,
  targetWidth: number,
  targetHeight: number,
  canvasWidth: number,
  canvasHeight: number
}
&nbsp;
export const validateCanvasExtractionValues = (extractionValues: ExtractionValues) =&gt; {
  // do not compute data if they're out of canvas area
  const {
    x,
    y,
    targetHeight,
    targetWidth,
    canvasHeight,
    canvasWidth
  } = extractionValues
  
  const newExtractionValues = { ...extractionValues }
  
  if (x &lt; 0) { newExtractionValues.x = 0 }
  if (y &lt; 0) { newExtractionValues.y = 0 }
  if (x + targetWidth &gt; canvasWidth) { newExtractionValues.targetWidth = canvasWidth - x }
  if (y + targetHeight &gt; canvasHeight) { newExtractionValues.targetHeight = canvasHeight - y }
  
  return newExtractionValues
}</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sun Feb 16 2020 15:30:19 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
