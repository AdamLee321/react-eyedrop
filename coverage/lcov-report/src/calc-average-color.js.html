<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/calc-average-color.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../index.html">All files</a> / <a href="index.html">src</a> calc-average-color.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>14/14</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>8/8</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>2/2</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>14/14</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">// @flow
&nbsp;
export const calcAverageColor = (colorBlock: Array&lt;{ r: number, g: number, b: number }&gt;): { r: number, g: number, b: number } =&gt; {
  const totalPixels = colorBlock.length
  
  if(typeof colorBlock !== 'object' || typeof colorBlock.reduce === 'undefined') {
    throw new Error('calcAverageColor: only accepts array of colors')
  }
  
  if(totalPixels === 0) {
    throw new Error('calcAverageColor: 0 pixels found')
  }
  
  if(totalPixels === 1) {
    return colorBlock[0]
  }
  
  const rgbAverage = colorBlock.reduce((rgbAcc, colorsObj) =&gt; {
    rgbAcc[0] += Math.round(colorsObj.r / totalPixels)
    rgbAcc[1] += Math.round(colorsObj.g / totalPixels)
    rgbAcc[2] += Math.round(colorsObj.b / totalPixels)
    return rgbAcc
  }, [0, 0, 0])
  
  return { r: rgbAverage[0], g: rgbAverage[1], b: rgbAverage[2] }
}</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sun Feb 16 2020 13:19:41 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../sorter.js"></script>
</body>
</html>
