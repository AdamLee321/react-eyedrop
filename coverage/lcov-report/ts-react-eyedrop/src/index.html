<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for ts-react-eyedrop/src</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> ts-react-eyedrop/src
    </h1>
    <div class='clearfix'>
    </div>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="calcAverageColor.ts"><a href="calcAverageColor.ts.html">calcAverageColor.ts</a></td>
	<td data-value="" class="pic high"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	</tr>

<tr>
	<td class="file high" data-value="getCanvasBlockColors.ts"><a href="getCanvasBlockColors.ts.html">getCanvasBlockColors.ts</a></td>
	<td data-value="" class="pic high"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	</tr>

<tr>
	<td class="file high" data-value="imageToCanvas.ts"><a href="imageToCanvas.ts.html">imageToCanvas.ts</a></td>
	<td data-value="" class="pic high"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct medium">%</td>
	<td data-value="" class="abs medium">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	</tr>

<tr>
	<td class="file high" data-value="parseRgb.ts"><a href="parseRgb.ts.html">parseRgb.ts</a></td>
	<td data-value="" class="pic high"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	</tr>

<tr>
	<td class="file high" data-value="rgbToHex.ts"><a href="rgbToHex.ts.html">rgbToHex.ts</a></td>
	<td data-value="" class="pic high"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	</tr>

<tr>
	<td class="file high" data-value="validateCanvasExtractionValues.ts"><a href="validateCanvasExtractionValues.ts.html">validateCanvasExtractionValues.ts</a></td>
	<td data-value="" class="pic high"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	</tr>

<tr>
	<td class="file high" data-value="validatePickRadius.ts"><a href="validatePickRadius.ts.html">validatePickRadius.ts</a></td>
	<td data-value="" class="pic high"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	<td data-value="" class="pct high">%</td>
	<td data-value="" class="abs high">/</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 24 2021 23:22:26 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
