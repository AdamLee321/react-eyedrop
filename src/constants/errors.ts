export const ERROR_MSGS = {
  calcAverageColor: {
    notArrayOfRgbObj: 'calcAverageColor: only accepts array of colors',
    noPixelsFound: 'calcAverageColor: 0 pixels found'
  },
  getCanvasBlockColors: {
    elementNotCanvas: 'getCanvasBlockColors: element is not of type canvas'
  },
  targetToCanvas: {
    targetNotElement: 'targetToCanvas: event target not HTML element'
  },
  imageToCanvas: {
    targetNotImg: 'imageToCanvas: event target not of node type img'
  },
  validatePickRadius: {
    pickRadiusOutOfBounds: 'Property `pickRadius` out of bounds: please choose a value between 0 - 450.'
  }
};

export const PICK_RADIUS_OUT_OF_BOUNDS_ERROR = new Error(ERROR_MSGS.validatePickRadius.pickRadiusOutOfBounds);

export const TARGET_NODE_TYPE_NOT_IMG_ERROR = new Error(ERROR_MSGS.imageToCanvas.targetNotImg);

export const TARGET_NOT_HTML_ELEMENT_ERROR = new Error(ERROR_MSGS.targetToCanvas.targetNotElement);

export const VAL_NOT_RGB_OBJ_ARRAY_ERROR = new Error(ERROR_MSGS.calcAverageColor.notArrayOfRgbObj);

export const ZERO_PIXELS_FOUND_ERROR = new Error(ERROR_MSGS.calcAverageColor.noPixelsFound);

export const ELEMENT_NOT_CANVAS_ERROR = new Error(ERROR_MSGS.getCanvasBlockColors.elementNotCanvas);
