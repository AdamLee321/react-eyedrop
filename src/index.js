// @flow
import React from 'react'
import type { Node } from 'react'
import html2canvas from 'html2canvas'
import getCanvasPixelColor from 'get-canvas-pixel-color'

import rgbToHex from './rgbToHex'

const styles = {
  eyedropperWrapper: {
    position: 'relative'
  },
  eyedropperWrapperButton: {
    backgroundColor: '#000000',
    color: '#ffffff',
    border: 'none',
    borderRadius: '20%',
    padding: '10px 25px',
  }
}

type Props = {
  onChange: Function,
  wrapperClasses?: string,
  buttonClasses?: string,
  customComponent?: Node,
  once?: boolean,
  cursorActive?: string,
  cursorInactive?: string,
  onInit?: Function,
  onPickStart?: Function,
  onPickEnd?: Function,
  passThrough?: string,
  pickRadius?: { unit: 'pixel' | 'radius', amount: number }
}

type State = {
  colors: {
    rgb: string,
    hex: string
  },
  pickingColorFromDocument: boolean
}

export default class EyeDropper extends React.Component<Props, State> {
  constructor(props) {
    super(props)
    this.state = {
      colors: { rgb: '', hex: '' },
      pickingColorFromDocument: false
    }
    this.cursorActive = props.cursorActive ? props.cursorActive : 'copy'
    this.cursorInactive = props.cursorInactive ? props.cursorInactive : 'auto'
  }

  componentDidMount() {
    const { onInit, pickRadius } = this.props
    if (onInit) { onInit() }

    if(pickRadius) {
     if (pickRadius.amount > 1){
       throw new Error('Amount should never be below 1.')
     }
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    if(nextProps.once !== this.props.once && nextProps.once) {
      document.removeEventListener('click', this.targetToCanvas)
      document.body.style.cursor = this.cursorInactive
      this.setState({ pickingColorFromDocument: false })
    }
  }

  pickColor = () => {
    const { onPickStart, once } = this.props
    const { pickingColorFromDocument } = this.state
    const { cursorActive } = this

    if (onPickStart) { onPickStart() }
    document.body.style.cursor = cursorActive
    if(!pickingColorFromDocument) {
      document.addEventListener('click', this.targetToCanvas, { once })
      this.setState({ pickingColorFromDocument: true })
    }
  }

  targetToCanvas = (e: *) => {
    const { target } = e
    const { pickRadius } = this.props

    html2canvas(target, { logging: false })
    .then((canvas) => {
      if (pickRadius) {
        this.extractColors(canvas, e)
      } else {
        this.extractColor(canvas, e)
      }
    })
  }

  extractColor = (canvas: *, e: *) => {
    const { offsetX, offsetY } = e

    const colors = getCanvasPixelColor(canvas, offsetX, offsetY)
    this.setColors(colors)
  }

  extractColors = (canvas: *, e: *) => {
    const { offsetX, offsetY } = e
    const { unit, amount } = this.props.pickRadius

    let maxRadius, minRadius
    if (unit === 'radius') {
      maxRadius = amount
      minRadius = -(amount) - 1
    } else if (unit === 'pixel') {
      if (amount % 2 !== 0) {
        maxRadius = ((amount - 1) / 2)
        minRadius = -((amount - 1) / 2) - 1
      } else {
        throw new Error('[EyeDrop] The unit \'pixel\' may only have an odd amount.')
      }
    } else {
      throw new Error('[EyeDrop] Please define a proper unit type.')
    }

    const colors = []
    let radialOffsetX, radialOffsetY

    for(let x = maxRadius; x !== minRadius; x--) {
      for(let y = maxRadius; y !== minRadius; y--) {
        radialOffsetX = (offsetX - x)
        radialOffsetY = (offsetY - y)

        if (!(radialOffsetX < 0) && !(radialOffsetY < 0)) {
          colors.push(getCanvasPixelColor(canvas, radialOffsetX, radialOffsetY))
        }
      }
    }
    this.calcAverageColor(colors)
  }

  calcAverageColor = (colors: Array<{r: number, g: number, b: number}>) => {
    let totalR = 0, totalG = 0, totalB = 0
    colors.map(({ r, g, b }, index) => {
      totalR += r * r
      totalG += g * g
      totalB += b * b
      if(index !== 0) {
        totalR = Math.sqrt(totalR / 2)
        totalG = Math.sqrt(totalG / 2)
        totalB = Math.sqrt(totalB / 2)
      }
    })
    const averageR = parseInt(totalR)
    const averageG = parseInt(totalG)
    const averageB = parseInt(totalB)
    this.setColors({ r: averageR, g: averageG, b: averageB })
  }

  setColors = ({ r, g, b }) => {
    const { cursorInactive } = this
    const { onPickEnd, passThrough, once } = this.props
    const rgb = `rgb(${r}, ${b}, ${g})`
    const hex = rgbToHex(r, b, g)

    if(once) { document.body.style.cursor = cursorInactive }
    if (passThrough) { this.setState({ colors: { rgb, hex } }) }
    this.props.onChange({ rgb, hex })
    if (onPickEnd) { onPickEnd() }
  }

  render() {
    const {
      wrapperClasses,
      buttonClasses,
      customComponent: CustomComponent,
      passThrough,
      children
    } = this.props

    const shouldPassThrough = passThrough ? { [passThrough]: this.state.colors } : {}

    return (
      <div style={styles.eyedropperWrapper} className={wrapperClasses}>
        {CustomComponent ? (
          <CustomComponent
            onClick={this.pickColor}
            {...shouldPassThrough}
          />
        ) : (
          <button
            style={styles.eyedropperWrapperButton}
            className={buttonClasses}
            onClick={this.pickColor}
          >
            {children ? children : 'Eye-Drop'}
          </button>
        )}
      </div>
    )
  }
}

EyeDropper.defaultProps = {
  once: true
}