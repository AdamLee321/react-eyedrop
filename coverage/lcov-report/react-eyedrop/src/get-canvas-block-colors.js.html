<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for react-eyedrop/src/get-canvas-block-colors.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">react-eyedrop/src</a> get-canvas-block-colors.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>11/11</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>2/2</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>1/1</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>11/11</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">225x</span>
<span class="cline-any cline-yes">225x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">// @flow
&nbsp;
import { ELEMENT_NOT_CANVAS_ERROR } from './constants/errors'
import type { RgbObj } from './types'
import { validateCanvasExtractionValues } from './validate-canvas-extraction-values'
&nbsp;
export const getCanvasBlockColors = (
  canvas: HTMLCanvasElement,
  x: number,
  y: number,
  width: number,
  height: number
): Array&lt;RgbObj&gt; =&gt; {
  if(!canvas.getContext) {
    throw ELEMENT_NOT_CANVAS_ERROR
  }
  const ctx = canvas.getContext('2d')
&nbsp;
  const validatedExtractionValues = validateCanvasExtractionValues({
    x,
    y,
    targetHeight: height,
    targetWidth: width,
    canvasHeight: canvas.height,
    canvasWidth: canvas.width
  })
  
  const imageData = ctx.getImageData(
    validatedExtractionValues.x,
    validatedExtractionValues.y,
    validatedExtractionValues.targetWidth,
    validatedExtractionValues.targetHeight
  ).data
&nbsp;
  let colorBlock = []
  for (let i = 0; i &lt; imageData.length; i += 4) {
    const [r, g, b, a] = imageData.slice(i, i + 4)
    colorBlock.push({ r: r, g: g, b: b })
  }
  return colorBlock
}
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sun Feb 16 2020 15:30:19 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
