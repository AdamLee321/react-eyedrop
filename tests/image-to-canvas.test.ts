import { ERROR_MSGS } from '../src/constants/errors'
import { imageToCanvas } from '../src/imageToCanvas'

describe('imageToCanvas should', () => {
  it('return a canvas based on image', () => {
    const imgElement = document.createElement('img')
    expect(imageToCanvas(imgElement)).toMatchSnapshot()
  })

  it('throw if event target is NOT of node type img', () => {
    const divElement = document.createElement('div')
    // @ts-ignore
    expect(() => imageToCanvas(divElement)).toThrow(ERROR_MSGS.imageToCanvas.targetNotImg)
  })
})
