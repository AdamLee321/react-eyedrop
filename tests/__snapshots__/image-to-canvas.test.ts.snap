// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imageToCanvas should return a canvas based on image 1`] = `Promise {}`;
